libsupport/include/katana/ProgressTracer.h

```diff
@@ -0,0 +1,222 @@
+#ifndef KATANA_LIBSUPPORT_KATANA_PROGRESSTRACER_H_
+#define KATANA_LIBSUPPORT_KATANA_PROGRESSTRACER_H_
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <variant>
+#include <vector>
+
+#include "katana/Result.h"
+#include "katana/config.h"
+
+/// This tracer does not currently support thread-local tracers or concurrency controls
+/// Functions should only be used in a single-threaded context
+///
+/// Best Practices:
+///   If possible always avoid creating raw ProgressSpans from ProgressTracer
+///   If possible always use ProgressScopes to handle ProgressSpans
+///   Only use 1 ProgressTracer in an execution, so they should be created at entry points
+///   Use ProgressScope's RAII to handle early returns (i.e. due to errors)
+///   Raw ProgressSpans should be used for special scenarios like tracing asynchronous calls
+///
+/// Notes:
+///   SharedMemSys and DistMemSys will initialize the global ProgressTracer to a JsonTracer
+///     later it will be set to a No-op Tracer, on Fini or the destructor they will call
+///     Close on the global ProgressTracer
+///   This prevents GetProgressTracer() from returning nullptr and ensures Tracers are closed
+///
+///   ProgressScope's will only close their ProgressSpan when their ProgressSpan is the active span
+///   This means that if a user exclusively uses ProgressScopes,
+///   Then parent ProgressSpans will not be closed until their child ProgressSpan's close
+///   i.e.
+///   {
+///      auto scope1 = tracer.StartActiveSpan("1");
+///      auto scope2 = tracer.StartActiveSpan("2");
+///      if (err) { return; }
+///      scope2.Close();
+///      auto scope3 = tracer.StartActiveSpan("3");
+///   }
+///   Always results in scope2's ProgressSpan finishing before scope1's
+///   and scope3's ProgressSpan finishing before scope1's if there is no error
```

Comment:
Couldn't parse this statement

Reply:
I'm trying to convey that you always get the behavior:

scope1 START
scope2 START
scope2 FINISH
scope1 FINISH

or

scope1 START
scope2 START
scope3 START
scope3 FINISH
scope2 FINISH
scope1 FINISH

depending on the error, as opposed to say scope1 finishing before scope2 so that scope2 is not contained entirely within the span of scope1

Reply:
Change to this maybe?
"Always scope2's ProgressSpan finish before scope1's ProgressSpan; 
 if there is no error, scope3's ProgressSpan finish before scope1's ProgressSpan"

Reply:
That's much better thanks